codec-redis/src/main/java/io/netty/handler/codec/redis/RedisEncoder.java

```diff
@@ -0,0 +1,150 @@
+        if (msg instanceof StringRedisMessage) {
+            writeInlineByteBufMessage(buf, (StringRedisMessage) msg);
+        } else if (msg instanceof IntegerRedisMessage) {
+            writeIntegerMessage(buf, (IntegerRedisMessage) msg);
+        } else if (msg instanceof BulkStringRedisMessage) {
+            writeBulkStringMessage(buf, (BulkStringRedisMessage) msg);
+        } else if (msg instanceof ArrayHeaderRedisMessage) {
+            writeArrayHeaderObject(buf, (ArrayHeaderRedisMessage) msg);
+        } else if (msg instanceof ArrayRedisMessage) {
+            writeArrayMessage(buf, (ArrayRedisMessage) msg);
+        } else {
+            throw new CodecException("bad message: " + msg);
+        }
+        return buf;
+    }
+
+    private static void writeInlineByteBufMessage(ByteBuf buf, StringRedisMessage msg) {
+        buf.writeByte(msg.type().value());
+        buf.writeBytes(msg.content().getBytes(CharsetUtil.UTF_8));
+        buf.writeBytes(CRLF);
+    }
+
+    private static void writeIntegerMessage(ByteBuf buf, IntegerRedisMessage msg) {
+        buf.writeByte(msg.type().value());
+        buf.writeBytes(numberToBytes(msg.value()));
+        buf.writeBytes(CRLF);
+    }
+
+    private static void writeBulkStringMessage(ByteBuf buf, BulkStringRedisMessage msg) {
+        buf.writeByte(msg.type().value());
+        if (msg.isNull()) {
+            buf.writeBytes(NULL_BULK_STRING);
+        } else {
+            ByteBuf content = msg.content();
+            buf.writeBytes(numberToBytes(content.readableBytes()));
+            buf.writeBytes(CRLF);
+            buf.writeBytes(content);
+        }
+        buf.writeBytes(CRLF);
+    }
+
+    /**
+     * Write array header only without body. Use this if you want to write arrays as streaming.
+     */
+    private static void writeArrayHeaderObject(ByteBuf buf, ArrayHeaderRedisMessage msg) {
+        buf.writeByte(RedisMessageType.ARRAY.value());
+        if (msg.isNull()) {
+            buf.writeBytes(NULL_BULK_STRING);
+        } else {
+            buf.writeBytes(numberToBytes(msg.length()));
+        }
+        buf.writeBytes(CRLF);
+    }
+
+    /**
+     * Write full constructed array message.
+     */
+    private static void writeArrayMessage(ByteBuf buf, ArrayRedisMessage msg) {
+        buf.writeByte(msg.type().value());
+        if (msg.isNull()) {
+            buf.writeBytes(NULL_BULK_STRING);
+            buf.writeBytes(CRLF);
+        } else {
+            buf.writeBytes(numberToBytes(msg.children().size()));
+            buf.writeBytes(CRLF);
+            for (RedisMessage child : msg.children()) {
+                writeRedisMessage(buf, child);
+            }
+        }
+    }
+
+    private static int calculateBufferLength(RedisMessage msg) {
+        if (msg instanceof StringRedisMessage) {
+            // TODO getBytes is not good..
+            return TYPE_LENGTH + ((StringRedisMessage) msg).content().getBytes(CharsetUtil.UTF_8).length + CRLF.length;
```

Comment:
I'm worrying about it. Redis server's responses are always US_ASCII. But, it's not a protocol's specification.